### PR TITLE
refactor(cluster): Move CLUSTER commands to their own methods.

### PR DESCRIPTION
**Pull request by chakaz**
While this change looks kinda big, it's pretty much all moved lines. I didn't modify anything beyond that.

### CODE REVIEW

File: src/server/server_family.cc [hunk 1]
```diff
@@ -1606,6 +1609,7 @@ Metrics ServerFamily::GetMetrics() const {
 }
 
 void ServerFamily::Info(CmdArgList args, ConnectionContext* cntx) {
+  // For more details https://redis.io/commands/cluster-slots/
```
Comment (adiholden) [inline src/server/server_family.cc:1612]: I dont think this comment belongs here

**Comment by adiholden** [comment]
@chakaz so I think now is a good time to move this functions to cluster family file if you think its needed

**Comment by chakaz** [comment]
Wow, that was kinda huge :)
In order to move all `CLUSTER` logic to its own files I had to do some refactors in a few places. Most notably, the `ServerFamily::replica_` field, which had its own mutex, now needs to be accessed from the new files as well. So I created a utility class to force access via locking (pretty neat if I may say so).
Otherwise, most changes are around cut-paste code, plumbing all pointers to their respected new places, etc.
I tested manually that all `CLUSTER` commands (like `CLUSTER SLOTS`, etc) still work. I also ran the cluster test under the `tests/` dir. I think it succeeded? (it says `8 passed` followed by a text that is also printed without my changes: `Task was destroyed but it is pending`).
Anyway, PTAL :)

File: src/server/CMakeLists.txt [hunk 2]
```diff
@@ -60,4 +61,4 @@ add_custom_target(check_dfly WORKING_DIRECTORY .. COMMAND ctest -L DFLY)
 add_dependencies(check_dfly dragonfly_test json_family_test list_family_test
                  generic_family_test memcache_parser_test rdb_test journal_test
                  redis_parser_test snapshot_test stream_family_test string_family_test
-                 bitops_family_test set_family_test zset_family_test hll_family_test cluster_config_test)
```
Comment (adiholden) [inline src/server/CMakeLists.txt:63]: why did you remove cluster_config_test?

File: src/server/cluster/cluster_family.cc [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+inline CommandId::Handler HandlerFunc(ClusterFamily* se, EngineFunc f) {
+  return [=](CmdArgList args, ConnectionContext* cntx) { return (se->*f)(args, cntx); };
+}
+
+#define HFUNC(x) SetHandler(HandlerFunc(this, &ClusterFamily::x))
+
+void ClusterFamily::Register(CommandRegistry* registry) {
+  *registry << CI{"CLUSTER", CO::READONLY, 2, 0, 0, 0}.HFUNC(Cluster)
+            << CI{"READONLY", CO::READONLY, 1, 0, 0, 0}.HFUNC(ReadOnly)
+            << CI{"READWRITE", CO::READONLY, 1, 0, 0, 0}.HFUNC(ReadWrite);
```
Comment (adiholden) [inline src/server/cluster/cluster_family.cc:267]: why did you add this commands in the PR? I didnt see they exist before

File: src/server/cluster/cluster_family.cc [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+                                   " 1 ", connect_state, " 0-16383\r\n");
+    if (!vec.empty()) {  // info about the replica
+      const auto& info = vec[0];
+      absl::StrAppend(&msg, etl.remote_client_id_, " ", info.address, ":", info.listening_port, "@",
+                      info.listening_port, " slave 0 ", server_family_->master_id(), " 1 ",
+                      connect_state, "\r\n");
+    }
+    return msg;
+  } else {
+    auto replica = server_family_->GetReplica();
```
Comment (adiholden) [inline src/server/cluster/cluster_family.cc:86]: The changes with MutexGuardedObject are because we use the info struct here?
If so I prefer having GetReplicaInfo function in server_family_

Comment (chakaz) [reply src/server/cluster/cluster_family.cc:86]: Yes, it's for info and for getting the replica's master.
I am happy to expose that function in `ServerFamily`, but I do think that it's safer like this (mutex is forced)

Comment (adiholden) [reply src/server/cluster/cluster_family.cc:86]: I would like this PR to be more simple for review. So if you think we need a change to enforce mutex lets discuss it and will decide if to create a different PR for this. But for now please just expose the ReplicaGetInfo and use the mutex as done before

Comment (chakaz) [reply src/server/cluster/cluster_family.cc]: Done.

File: src/server/cluster/cluster_family.cc [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+    (*cntx)->SendLong(0);                           // start sharding range
+    (*cntx)->SendLong(ClusterConfig::kMaxSlotNum);  // end sharding range
+    BuildClusterSlotNetworkInfo(cntx, preferred_endpoint, absl::GetFlag(FLAGS_port),
+                                server_family_->master_id());
+    if (!vec.empty()) {  // info about the replica
+      const auto& info = vec[0];
+      BuildClusterSlotNetworkInfo(cntx, info.address, info.listening_port, etl.remote_client_id_);
+    }
+  } else {
+    Replica::Info info = server_family_->GetReplicaInfo();
```
Comment (adiholden) [inline src/server/cluster/cluster_family.cc:155]: There is a bug here that if somesone run replicaof no one while this command is running you can get a segfault here as GetReplicaInfo accesses the replica_ ptr without checking its not null.
I believe this bug was also before this PR. 
Please fix this in another PR, not this one

**Review by adiholden — APPROVED** [review]

File: src/server/cluster/cluster_family.cc (hunk 5 above)
Comment (chakaz) [reply src/server/cluster/cluster_family.cc:155]: Sounds like a job for `MutexGuardedObject`! :)
As you said, that can be in another PR though. And there might be a simpler way too :)
